common/copr_common/subprocess_live_output.py

```diff
@@ -0,0 +1,169 @@
+"""
+Streamed reading of Popen() stdout and stderr:
+
+    proc = PosixPipedProcess(["command"])
+    for chunk, output = proc.readchunks()
+        assert output in ["stdout", "stderr"]
+        process(chunk)
+"""
+
+from subprocess import Popen, PIPE
+from threading import Thread
+from queue import Queue
+import select
+import time
+
+EOF = 0
+CUT = 1
+LOOP = 2
+TIMEOUT = 3
+ERR = -1
+
+def _pipe_consumer_thread(pipe, queue, identifier, maxlen=None, poll=5.0):
+    """
+    Read the PIPE in a separate thread, and enqueue the chunks of output into
+    the QUEUE.  The chunks are stored as:
+        [(pipename, chunk), ...,  # normal chunks (of bytes)
+         (pipename, EOF|CUT|ERR), # pre-termination sequence
+         None,                    # iter(sentinel) (stop iteration)
+        ]
+    """
+    counter = 0
+    fd = pipe.fileno()
+    try:
+        while True:
+            readable, _, exc = select.select([fd], [], [fd], poll)
+            if exc:
+                queue.put((identifier, ERR))
+                break
+
+            if not readable:
+                queue.put((identifier, LOOP))
+                continue
+
+            chunk = pipe.read1(1024)
```

Comment:
is this typo? I haven't find anything about `read1`, does it even exist?

Reply:
Yes, read() waits till the buffer is filled, or EOF. We need to process partial/incomplete reads.